### DISCUSSION TITLE
Getting hovered states after Render()

### DISCUSSION THREAD

**Discussion by WickedP666** [ocornut/imgui#7279] [New Users Build/Link/Run issues ONLY!]
Hi folks,

I have an off-screen GLFW window, used as an off-screen renderer. I'm catching the rendered image and displaying it in another dialog, outside of GLFW. Unfortunately, I have to do it this way, because the SDK I'm using does not communicate with windows other than it's own. The software is a plugin that works inside another well-known 3D application, which has it's own dialog system. It only communicates with it's own dialogs. Everything, including the mouse and user interaction catches, are handled in the main (non-GLFW) dialog.

I came across Dear Imgui recently in the hope it would replace my own custom GUI interface engine. ImGui seems to have provided a really good and helpful alternative on this front:

![image](https://github.com/ocornut/imgui/assets/156973885/826b5058-76dc-4133-b1f0-68bda7065037)

However, because of the way my dialog system has to work, I need to be able to execute things like mouse testing after the call to Render(). Is it possible for example, to manually test some sort of window/item map to see what's under the mouse at any position on the last frame?

I've made my own wrapper to handle the drawing and can change the styles to handle my own focusing etc. So I'm prepared to do some manual work to handle the 'interactive' things.

I've spotted functions like ImGui::UpdateHoveredWindowAndCaptureFlags(), but these don't appear to cover my needs (happy to be corrected). It seems as though things need to be set in ImGui first, before being rendered. Whereas I need to render, then test to see what's there.

I realise this may not quite be ImGui's intended way of use here (again, happy to be educated), but is there an alternative method that could make this work? I'm not quite sure of the solution, if there is one. An ID map I can read myself maybe? What would/does this look like?

Hoping I've explained enough for context. Unfortunately my hands are tied on the windowing system and the process order. And apologies if a solution already exists, I just can't see it or make it work. If you need more from me, let me know.

Regards,

WP.

### For reference:

Version:
#define IMGUI_VERSION       "1.90.2 WIP"
#define IMGUI_VERSION_NUM   19013

Files:
#include "imgui-master\imgui.h"
#include "imgui-master\imgui_internal.h"
#include "imgui-master\misc\cpp\imgui_stdlib.h"
#include "imgui-master\misc\single_file\imgui_single_file.h"
//#include "imgui-master\misc\freetype\imgui_freetype.h"
#include "imgui-master\backends\imgui_impl_glfw.h"
#include "imgui-master\backends\imgui_impl_opengl3.h"

**Comment by ocornut** [comment]
Hello,
This is not the right section to ask.

Beside, you should specify why you are doing this, e.g.

> Is it possible for example, to manually test some sort of window/item map to see what's under the mouse at any position on the last frame?

Why would you need that after Render() ? Why can't you query the imgui state you need during the imgui frame etc. The issue has nothing to do with rendering, since imgui doesn't care about when you are rendering.

**Reply by WickedP666** [reply]
Thanks ocornut,

Is there a better place to ask? Happy to redirect.

In regards to this query, I realise it isn't to do with rendering itself, but the order of rendering matters for my application. There's several draw stages, including CPU-bound ones. I spent some time trying to rework the windowing system and the draw pipeline, and have managed to get it drawing 'successfully'. I think it's enough for me to work around for now.

I have another query regarding InputInt() and the + - button ids. Happy to direct that somewhere else (the "issues" tab?).

WP.

**Reply by ocornut** [reply]
> Is there a better place to ask? 

You are supposed to open an issue, it appears that to be spelled out everywhere in this section :)

When you do the rendering doesn’t matter to imgui.
But querying hovered state has nothing to do with rendering.
So you will need to clarify your situation and what you need to do with that info to find out what info you actually need, and verify you are not trying to reinvent a broken version of io.WantCaptureMouse (see FAQ) at this is a very common mistake.
